### PR TITLE
Reload channels and tune async connection

### LINKED ISSUE #202: Channel Service Refs Can sometimes have alphabetic characters appended at the end
Create a function to remove the characters so that sref is consistent across E2 data types and APIs.

### PR DESCRIPTION
**Pull request by phunkyfish**
v3.20.0
- Added: Add options to Tune Async connection for slower E2 devices
- Added: Enable option to reload channels and groups once per day if changes are detected
- Added: Allow timers generated by autotimers to be deleted
- Fixed: Fix for disabling timers
- Added: Use genre from kodi for timers instead of lookup where possible
- Added: Support recording type fully for TV/Radio/Any channel
- Added: By default load Last Scanned group if no TV groups found
- Added: Limit autotimers to TV/Radio or originating channel's groups
- Fixed: Channel Service Refs Can sometimes have alphabetic characters appended at the end, fixes #202
- Added: For timers and autotimers without a valid channel put in error state instead of omitting
- Fixed: IsRunning timer check not checking start time

### CODE REVIEW

**Comment by phunkyfish** [comment]
Pushing small change to second from last commit

**Comment by phunkyfish** [comment]
I might add an additional small feature later on if you are not in the process of reviewing.

**Comment by phunkyfish** [comment]
I’ll add the fixes/workarounds I found related to any channel at some point today.

**Comment by phunkyfish** [comment]
@ksooo quite a few fixes came from the any channel kodi issue. Should be ready now.

File: src/enigma2/ConnectionManager.cpp [hunk 1]
```diff
@@ -140,7 +140,9 @@ void* ConnectionManager::Process()
 {
   static bool log = false;
   static unsigned int retryAttempt = 0;
-  const Settings &settings = Settings::GetInstance();
+//  const Settings &settings = Settings::GetInstance();
```
Comment (ksooo) [inline src/enigma2/ConnectionManager.cpp]: remove this?

File: src/enigma2/Timers.cpp [hunk 2]
```diff
@@ -776,13 +798,41 @@ PVR_ERROR Timers::UpdateAutoTimer(const PVR_TIMER &timer)
 
     if (timerToUpdate.GetAnyChannel() && timer.iClientChannelUid != PVR_TIMER_ANY_CHANNEL)
     {
+      std::string serviceReference = m_channels.GetChannel(timer.iClientChannelUid)->GetServiceReference();
```
Comment (ksooo) [inline src/enigma2/Timers.cpp]: const?

File: src/enigma2/data/Channel.cpp [hunk 3]
```diff
@@ -53,6 +53,9 @@ bool Channel::UpdateFrom(TiXmlElement* channelNode)
   if (!XMLUtils::GetString(channelNode, "e2servicename", m_channelName))
     return false;
 
+  m_fuzzyChannelName = m_channelName;
+  m_fuzzyChannelName.erase(remove_if(m_fuzzyChannelName.begin(), m_fuzzyChannelName.end(), isspace), m_fuzzyChannelName.end());
```
Comment (ksooo) [inline src/enigma2/data/Channel.cpp]: std::remove_if? why does this compile? Any evil `using namespace std` left somewhere?

File: src/enigma2/data/RecordingEntry.h [hunk 4]
```diff
@@ -58,27 +58,42 @@ namespace enigma2
...
       bool UpdateFrom(TiXmlElement* recordingNode, const std::string &directory, enigma2::Channels &channels);
       void UpdateTo(PVR_RECORDING &left, Channels &channels, bool isInRecordingFolder);
 
     private:
       long TimeStringToSeconds(const std::string &timeString);
 
+      std::shared_ptr<Channel> FindChannel(enigma2::Channels &channels);
+      std::shared_ptr<Channel> GetChannelFromChannelReferenceTag(enigma2::Channels &channels);
+      std::shared_ptr<Channel> GetChannelFromChannelNameSearch(enigma2::Channels &channels);
+      std::shared_ptr<Channel> GetChannelFromChannelNameFuzzySearch(enigma2::Channels &channels);
```
Comment (ksooo) [inline src/enigma2/data/RecordingEntry.h]: make all 4 new methods const?

Comment (ksooo) [reply src/enigma2/data/RecordingEntry.h]: ... and make `m_anyChannelTimerSource`, `m_radio` and `m_haveChannelType` `mutable `  => logical constness of `FindChannel`

File: src/enigma2/data/Tags.h [hunk 5]
```diff
@@ -48,7 +53,7 @@ namespace enigma2
         return (regex_match(m_tags, regex));
       }
 
-      void AddTag(const std::string &tagName, const std::string &tagValue = "")
+      void AddTag(const std::string &tagName, std::string tagValue = "", bool replaceUnderscores = false)
```
Comment (ksooo) [inline src/enigma2/data/Tags.h]: why the change from const std::string& to `by value` parameter?

File: src/enigma2/Epg.h [hunk 6]
```diff
@@ -54,6 +54,7 @@ namespace enigma2
     std::string LoadEPGEntryShortDescription(const std::string &serviceReference, unsigned int epgUid);
     data::EpgPartialEntry LoadEPGEntryPartialDetails(const std::string &serviceReference, time_t startTime);
     data::EpgPartialEntry LoadEPGEntryPartialDetails(const std::string &serviceReference, unsigned int epgUid);
+    std::string FindServiceReference(const std::string &title, int epgUid, time_t startTime, time_t endTime);
```
Comment (ksooo) [inline src/enigma2/Epg.h]: make this method const?

File: src/enigma2/data/RecordingEntry.h [hunk 7]
```diff
@@ -58,27 +58,42 @@ namespace enigma2
...
+
       std::string m_recordingId;
       time_t m_startTime;
       int m_duration;
       int m_lastPlayedPosition;
       std::string m_streamURL;
       std::string m_edlURL;
-      std::string m_channelName;
+      std::string m_channelName = "";
```
Comment (ksooo) [inline src/enigma2/data/RecordingEntry.h]: no need to init explicitely. string default ctor will be called automatically.

File: src/enigma2/data/RecordingEntry.h [hunk 8]
```diff
@@ -58,27 +58,42 @@ namespace enigma2
...
+      int m_channelUniqueId = PVR_CHANNEL_INVALID_UID;
       std::string m_directory;
-      std::string m_iconPath;
+      std::string m_iconPath = "";
```
Comment (ksooo) [inline src/enigma2/data/RecordingEntry.h]: no need to init explicitely. string default ctor will be called automatically

File: src/enigma2/data/AutoTimer.cpp [hunk 9]
```diff
@@ -158,13 +160,18 @@ bool AutoTimer::UpdateFrom(TiXmlElement* autoTimerNode, Channels &channels)
         m_channelId = channels.GetChannelUniqueId(Channel::NormaliseServiceReference(strTmp.c_str()));
 
         // Skip autotimers for channels we don't know about, such as when the addon only uses one bouquet or an old channel referene that doesn't exist
-        if (m_channelId < 0)
+        if (m_channelId == PVR_CHANNEL_INVALID_UID)
         {
-          Logger::Log(LEVEL_DEBUG, "%s could not find channel so skipping autotimer: '%s'", __FUNCTION__, m_title.c_str());
-          return false;
+          // Logger::Log(LEVEL_DEBUG, "%s could not find channel so skipping autotimer: '%s'", __FUNCTION__, m_title.c_str());
+          // return false;
```
Comment (ksooo) [inline src/enigma2/data/AutoTimer.cpp]: I would remove this two commented lines.

File: src/enigma2/data/RecordingEntry.h (hunk 4 above)
Comment (phunkyfish) [reply src/enigma2/data/RecordingEntry.h]: Is mutable a requirement of making the methods const?

**Comment by phunkyfish** [comment]
Completely agree on the settings, will look at adding levels to the settings. I.e. normal, advanced, expert. The main issue is Enigma2 has so many versions and images it’s nigh impossible to cater for every situation without them. That being said it’s a smaller percentage of the population require them so the levels of possible should help.

File: src/enigma2/data/RecordingEntry.h (hunk 4 above)
Comment (ksooo) [reply src/enigma2/data/RecordingEntry.h]: As FindChannel writes to these members, yes. But there is nothing bad with this. That's what mutable is for - achieve logical constness of a method.

Comment (phunkyfish) [reply src/enigma2/data/RecordingEntry.h]: Very cool, probably have a few more places I could use this!

File: src/enigma2/ConnectionManager.cpp (hunk 1 above)
Comment (phunkyfish) [reply src/enigma2/ConnectionManager.cpp]: Done

File: src/enigma2/Timers.cpp (hunk 2 above)
Comment (phunkyfish) [reply src/enigma2/Timers.cpp]: Already done

File: src/enigma2/data/Channel.cpp (hunk 3 above)
Comment (phunkyfish) [reply src/enigma2/data/Channel.cpp]: I have no idea. But there is no `using namespace std` anywhere.

File: src/enigma2/data/Tags.h (hunk 5 above)
Comment (phunkyfish) [reply src/enigma2/data/Tags.h]: Because of the default value.

Comment (ksooo) [reply src/enigma2/data/Tags.h]: imo a default value `const std::string& foo = ""`does work just fine.

File: src/enigma2/Epg.h (hunk 6 above)
Comment (phunkyfish) [reply src/enigma2/Epg.h]: Done

File: src/enigma2/data/RecordingEntry.h (hunk 7 above)
Comment (phunkyfish) [reply src/enigma2/data/RecordingEntry.h]: done

File: src/enigma2/data/RecordingEntry.h (hunk 8 above)
Comment (phunkyfish) [reply src/enigma2/data/RecordingEntry.h]: Done

File: src/enigma2/data/AutoTimer.cpp (hunk 9 above)
Comment (phunkyfish) [reply src/enigma2/data/AutoTimer.cpp]: Done

File: src/enigma2/data/Tags.h (hunk 5 above)
Comment (phunkyfish) [reply src/enigma2/data/Tags.h]: So it does ;)

I passed by value instead of creating a new variable in the method to be used here:

`std::replace(tagValue.begin(), tagValue.end(), ' ', '_');`

Is it better to pass const and make a copy?

File: src/enigma2/data/RecordingEntry.h (hunk 4 above)
Comment (phunkyfish) [reply src/enigma2/data/RecordingEntry.h]: Done

**Comment by phunkyfish** [comment]
Should be good now

**Comment by phunkyfish** [comment]
So I looked at the settings levels and they appear to be supported through the addon settings UI, however it doesn't appear to be possible to change the level. Neither can you add help dialogs. Bummer

**Comment by phunkyfish** [comment]
@ksooo does the PVR component of kodi manage the settings dialogs for the addons? If so it should be possible to support both levels for the settings and the help info. If the complexity to add this isn't too high I could try to add support for it if you think it's worthwhile?

**Comment by ksooo** [comment]
> does the PVR component of kodi manage the settings dialogs for the addons?

 no, its generic functionality for all (binary) addons.

> however it doesn't appear to be possible to change the level. 

 this is by design, afaik. there is only one application-wide level.  so, it feels okay that an addon cannot change the level.

> Neither can you add help info

 missing feature or bug? don't know.

> if you think it's worthwhile?

 not really. :-)

File: src/enigma2/data/Tags.h (hunk 5 above)
Comment (ksooo) [reply src/enigma2/data/Tags.h]: Your decision. Doesn't really make a difference, I'd say.

Comment (ksooo) [reply src/enigma2/data/Tags.h]: I'd go with the const ref way + copy in function body. So, it is very clear what happens. Performance-wise it does not make a difference, as both approaches need one copy.

Comment (phunkyfish) [reply src/enigma2/data/Tags.h]: Ok, cool will push an update for that later today.

**Comment by phunkyfish** [comment]
I’ll file a bug for the binary addon help info later and we’ll find out 😉

File: src/enigma2/data/Tags.h (hunk 5 above)
Comment (ksooo) [reply src/enigma2/data/Tags.h]: Your decision. From my pow the PR can go in as it is now.

Comment (phunkyfish) [reply src/enigma2/data/Tags.h]: Then in it goes...
